source/How-To-Guides/Use-Cobra.rst

```diff
@@ -1,7 +1,351 @@
+                            directives. This option will take precedence over the --include_dirs
+                            options and any directories specified using --include_dirs will be ignored.
+                            Instead, ament_cobra will gather all preprocessor options from the
+                            compile_commands.json file. (default: None)
+    --xunit-file XUNIT_FILE
+                            Generate a xunit compliant XML file (default: None)
+    --sarif-file SARIF_FILE
+                            Generate a SARIF file (default: None)
+    --cobra-version       Get the cobra version, print it, and then exit (default: False)
+    --verbose             Display verbose output (default: False)
+
+Basic operation
+---------------
+
+The ``ament_cobra`` tool:
+
+* Gathers a list of C/C++ filenames to be analyzed from from a list directories and/or files specified on the command line. Each directory is recursively scanned for target files. By default, if no filenames or directories are specified, ``ament_cobra`` scans for target files starting in the current working directory.
+* Analyzes the files using using one of a set of available rule sets (by default, ``ament_cobra`` uses Cobra's ``basic`` rule set)
+* Saves detailed results to a file, the name of which is based on the ruleset used (``_Basic_.txt``, for example)
+* Writes a summary of the issues discovered to stdout
+
+For example, the following invocation of ``ament_cobra``, without any command-line arguments,
+
+.. code-block::
+
+    spaceros-user@c1b264aeb129:~/spaceros/src/rcutils/src$ ament_cobra
+
+scans for all C and C++ files in the current working directory (``~/spaceros/src/rcutils/src``) and below, analyzes those files using Cobra's Basic rule set, and produces the following output:
+
+.. code-block::
+
+    1: Shadow, 9 patterns :: Fct names also used as variables
+    2: Rval, 1 patterns :: Inconsistent checking return value of function
+    3: Ifelseif, 9 patterns :: Missing else at end of if-else-if chain
+    4: SideEffect, 3 patterns :: More than one side-effect in expression
+    5: Macros, 46 patterns :: Macro argument not enclosed in parentheses
+    detailed results are in file _Basic_.txt
+
+    68 errors
+
+The output includes a summary of the issues found and the number of occurances of each issue.
+Cobra also saves a detailed summary of the analysis results in the file ``_Basic_.txt``, which is a JSON file that contains an entry for each issue.
+For example, the first couple of entries in ``__Basic__.txt`` for this run are as follows:
+
+.. code-block::
+
+    [
+    { "type"      :       "Fct names also used as variables",
+        "message"   :       "lines 688..688",
+        "file"      :       "/home/spaceros-user/spaceros/src/rcutils/src/logging.c",
+        "line"      :       688,
+        "cobra"     :       "1 10 0"
+    }
+    { "type"      :       "Fct names also used as variables",
```

Comment:
Is this file a single JSON array, in which case there should probably be a comma here, or a sequence of individual JSON objects, in which case the opening square bracket is probably incorrect?

Reply:
I just checked the nightly Docker image and cobra outputs this file without the comma (an invalid JSON format). It's a bit weird as the file output by cobra has a .txt filename extention, but it's supposed to be a JSON file. The .txt file is then read by their json_convert utility to generate SARIF. The utility reads the file OK and generates valid SARIF. 

So, I can think of a couple options:
* Use the term "JSON-like" in the docs and leave the doc as-is
* Update the doc to include the comma and enter an issue on the Cobra repo

Thoughts?

Reply:
> Thoughts?

Oof.

So a "stream" of JSON documents is equally valid contents. But I think if the file contains opening and closing square brackets / braces (`[]`) but does not put commas between objects then it is neither a stream of multiple JSON documents nor a single valid JSON array and that's probably worth ticketing upstream.

In the meantime, our documentation should accurately reflect the current reality, whether that be a single JSON document, a sequence of them, or some almost-JSON format.

Reply:
Ok, thanks. I'll ticket that upstream and keep the docs as-is to reflect the current state of things. Maybe introduce "JSON-like" at that point in the doc.